packages/vite/src/module-runner/hmrHandler.ts

```diff
@@ -123,3 +121,46 @@ class Queue {
     return true
   }
 }
+
+function getModulesByFile(runner: ModuleRunner, file: string) {
+  const modules: string[] = []
+  for (const [id, mod] of runner.moduleCache.entries()) {
+    if (mod.meta && 'file' in mod.meta && mod.meta.file === file) {
+      modules.push(id)
+    }
+  }
+  return modules
+}
+
+function getModulesEntrypoints(
+  runner: ModuleRunner,
+  modules: string[],
+  visited = new Set<string>(),
+  entrypoints = new Set<string>(),
+) {
+  for (const moduleId of modules) {
+    if (visited.has(moduleId)) continue
+    visited.add(moduleId)
+    const module = runner.moduleCache.getByModuleId(moduleId)
+    if (module.importers && !module.importers.size) {
+      entrypoints.add(moduleId)
+      continue
+    }
+    for (const importer of module.importers || []) {
+      getModulesEntrypoints(runner, [importer], visited, entrypoints)
+    }
+  }
+  return entrypoints
+}
+
+function findAllEntrypoints(
+  runner: ModuleRunner,
+  entrypoints = new Set<string>(),
+): Set<string> {
+  for (const [id, mod] of runner.moduleCache.entries()) {
+    if (mod.importers && !mod.importers.size) {
+      entrypoints.add(id)
+    }
+  }
+  return entrypoints
+}
```

Comment:
Could this miss entry points like:
```js
const { a } = runner.import('a.js')
const { b } = runner.import('b.js')
```
where `b.js` imports `a.js`? Maybe all modules that where called directly by `import` should be marked as entries? Note: maybe I'm misunderstanding how this works.

Reply:
Well, yes, but then it's not an entry point. In your example, we would call `a` twice where the second call will be cached anyway

Reply:
Ok, so when `b.js` is modified, then things will work well then.